sql/moz-fx-data-shared-prod/subscription_platform_derived/stripe_subscriptions_history_v1/query.sql

```diff
@@ -200,44 +200,114 @@ plans AS (
-    customers.shipping_address_country,
+    COALESCE(
+      NULLIF(customers.address_country, ""),
+      pre_fivetran_customers.address_country
+    ) AS address_country,
+    COALESCE(
+      IF(LENGTH(customers.address_state) = 2, customers.address_state, NULL),
+      pre_fivetran_customers.address_state
+    ) AS address_state,
+    NULLIF(customers.shipping_address_country, "") AS shipping_address_country,
+    COALESCE(
+      NULLIF(customers.shipping_address_state, ""),
+      us_shipping_zip_codes.state_code,
+      ca_shipping_postal_districts.province_code
+    ) AS shipping_address_state,
   FROM
     `moz-fx-data-shared-prod`.stripe_external.customer_v1 AS customers
   FULL JOIN
     -- Include customers that were deleted before the initial Fivetran Stripe import.
     `moz-fx-data-shared-prod`.stripe_external.pre_fivetran_customers_v1 AS pre_fivetran_customers
   USING
     (id)
+  LEFT JOIN
+    `bigquery-public-data.geo_us_boundaries.zip_codes` AS us_shipping_zip_codes
+  ON
+    customers.shipping_address_country = "US"
+    AND LEFT(customers.shipping_address_postal_code, 5) = us_shipping_zip_codes.zip_code
+  LEFT JOIN
+    ca_postal_districts AS ca_shipping_postal_districts
+  ON
+    customers.shipping_address_country = "CA"
+    AND LEFT(customers.shipping_address_postal_code, 1) = ca_shipping_postal_districts.district
 ),
 charges AS (
   SELECT
     charges.id AS charge_id,
-    COALESCE(cards.country, charges.billing_detail_address_country) AS country,
+    COALESCE(NULLIF(charges.billing_detail_address_country, ""), cards.country) AS country,
+    COALESCE(
+      NULLIF(charges.billing_detail_address_state, ""),
+      us_zip_codes.state_code,
+      ca_postal_districts.province_code
+    ) AS state,
   FROM
     `moz-fx-data-shared-prod`.stripe_external.charge_v1 AS charges
   JOIN
     `moz-fx-data-shared-prod`.stripe_external.card_v1 AS cards
   ON
     charges.card_id = cards.id
+  LEFT JOIN
+    `bigquery-public-data.geo_us_boundaries.zip_codes` AS us_zip_codes
+  ON
+    COALESCE(NULLIF(charges.billing_detail_address_country, ""), cards.country) = "US"
+    AND LEFT(charges.billing_detail_address_postal_code, 5) = us_zip_codes.zip_code
+  LEFT JOIN
+    ca_postal_districts
+  ON
+    COALESCE(NULLIF(charges.billing_detail_address_country, ""), cards.country) = "CA"
+    AND UPPER(LEFT(charges.billing_detail_address_postal_code, 1)) = ca_postal_districts.district
   WHERE
     charges.status = "succeeded"
 ),
-invoices_provider_country AS (
+invoices_provider_location AS (
   SELECT
     invoices.subscription_id,
     IF(
       JSON_VALUE(invoices.metadata, "$.paypalTransactionId") IS NOT NULL,
-      -- FxA copies PayPal billing agreement country to customer address.
-      STRUCT("Paypal" AS provider, customers.address_country AS country),
-      ("Stripe", charges.country)
+      -- FxA copied PayPal billing address to customer address before we enabled Stripe Tax (FXA-5457).
```

Comment:
It was news to me that FxA was no longer copying PayPal billing addresses to customer billing addresses (99.8% of recent customers have no billing address), and that's the reason why I avoided using the customer billing addresses as a resource for doing the postal-code-to-state mapping.